formats/json/commonMain/src/kotlinx/serialization/json/internal/StringOps.kt

```diff
@@ -71,3 +71,32 @@ internal fun String.toBooleanStrictOrNull(): Boolean? = when {
     this.equals("false", ignoreCase = true) -> false
     else -> null
 }
+
+private val INT_NUMBERS_CHARS = arrayOf('1', '2', '3', '4', '5', '6', '7', '8', '9', '0', 'e', 'E', '+', '-')
+internal fun String.toLongJson(): Long {
+    return toLongOrNull() ?: toDouble()
+        .takeIf { all { it in INT_NUMBERS_CHARS } }
+        ?.takeIf { it < Long.MAX_VALUE && it > Long.MIN_VALUE }
+        ?.toLong()
+    ?: throw NumberFormatException("Invalid number format: $this")
+}
+internal fun String.toLongJsonOrNull(): Long? {
+    return toLongOrNull() ?: toDoubleOrNull()
+        ?.takeIf { all { it in INT_NUMBERS_CHARS } }
```

Comment:
I think such code poses hidden problem: `all` iterates all symbols and `it in` iterates all chars in the `INT_NUMBERS_CHARS` array, so it results in a quadratic complexity. Moreover, it is just hard to read. I see several possibilities here:

1. To simplify condition, I think it is possible just to check `'.' !in it`. All other non-digit and non-E signs should be prohibited by `toDoubleOrNull()`, I guess?
2. A harder, but more correct change would be extract parsing logic from `consumeNumericLiteral` or to create `StringJsonLexer` in place so we wouldn't have two different code paths doing the same thing. 

I think it is possible to also extract these `takeIf`s to a separate function since it doesn't matter if you use `toDouble` or `toDoubleOrNull` given that exception is anyway thrown at the end if needed.

Reply:
I think  `StringJsonLexer` is the best solution